dump/dump.c

```diff
@@ -0,0 +1,248 @@
+		unsigned long long bitmap_len)
+{
+	unsigned int count = 0;
+	unsigned long long i;
+
+	for (i = 0; i < bitmap_len; i++)
+		count += used_bit[bitmap[i]];
+
+	return count;
+}
+
+int exfat_show_ondisk_all_info(struct exfat_blk_dev *bd)
+{
+	struct pbr *ppbr;
+	struct bsx64 *pbsx;
+	struct exfat_dentry *ed;
+	unsigned int root_clu_off, bitmap_clu_off, bitmap_clu;
+	unsigned int total_clus, used_clus, clu_offset, root_clu;
+	unsigned long long bitmap_len;
+	int ret;
+	unsigned char *bitmap;
+	char *volume_label;
+
+	ppbr = malloc(bd->sector_size);
+	if (!ppbr) {
+		exfat_err("Cannot allocate pbr: out of memory\n");
+		return -1;
+	}
+
+	/* read main boot sector */
+	ret = exfat_read_sector(bd, (char *)ppbr, BOOT_SEC_IDX);
+	if (ret < 0) {
+		exfat_err("main boot sector read failed\n");
+		ret = -1;
+		goto free_ppbr;
+	}
+
+	pbsx = &ppbr->bsx;
+
+	if (pbsx->sect_size_bits < EXFAT_MIN_SECT_SIZE_BITS ||
+	    pbsx->sect_size_bits > EXFAT_MAX_SECT_SIZE_BITS) {
+		exfat_err("bogus sector size bits : %u\n",
+				pbsx->sect_size_bits);
+		return -EINVAL;
+	}
+
+	if (pbsx->sect_per_clus_bits > 25 - pbsx->sect_size_bits) {
+		exfat_err("bogus sectors bits per cluster : %u\n",
+				pbsx->sect_per_clus_bits);
+		return -EINVAL;
+	}
+
+	if (bd->sector_size != 1 << pbsx->sect_size_bits) {
+		exfat_err("bogus sectors size : %u(sector size bits : %u)\n",
+				bd->sector_size, pbsx->sect_size_bits);
+
+	}
+
+	clu_offset = le32_to_cpu(pbsx->clu_offset);
+	total_clus = le32_to_cpu(pbsx->clu_count);
+	root_clu = le32_to_cpu(pbsx->root_cluster);
+
+	exfat_info("-------------- Dump Boot sector region --------------\n");
+	exfat_info("Volume Length(sectors): \t\t%" PRIu64 "\n",
+			le64_to_cpu(pbsx->vol_length));
+	exfat_info("FAT Offset(sector offset): \t\t%u\n",
+			le32_to_cpu(pbsx->fat_offset));
+	exfat_info("FAT Length(sectors): \t\t\t%u\n",
+			le32_to_cpu(pbsx->fat_length));
+	exfat_info("Cluster Heap Offset (sector offset): \t%u\n", clu_offset);
+	exfat_info("Cluster Count: \t\t\t\t%u\n", total_clus);
+	exfat_info("Root Cluster (cluster offset): \t\t%u\n", root_clu);
+	exfat_info("Volume Serial: \t\t\t\t0x%x\n", le32_to_cpu(pbsx->vol_serial));
+	exfat_info("Sector Size Bits: \t\t\t%u\n", pbsx->sect_size_bits);
+	exfat_info("Sector per Cluster bits: \t\t%u\n\n", pbsx->sect_per_clus_bits);
+
+	bd->cluster_size =
+		1 << (pbsx->sect_per_clus_bits + pbsx->sect_size_bits);
+	root_clu_off = exfat_calc_cluster_offset(bd, clu_offset, root_clu);
+
+	ed = malloc(sizeof(struct exfat_dentry)*3);
+	if (!ed) {
+		exfat_err("failed to allocate memory\n");
+		ret = -ENOMEM;
+		goto free_ppbr;
+	}
+
+	ret = exfat_read(bd->dev_fd, ed, sizeof(struct exfat_dentry)*3,
+			root_clu_off);
+	if (ret < 0) {
+		exfat_err("bitmap entry read failed: %d\n", errno);
+		ret = -EIO;
+		goto free_entry;
+	}
+
+	volume_label = exfat_conv_volume_serial(&ed[0]);
+	if (!volume_label)
+		return -EINVAL;
+
+	bitmap_clu = le32_to_cpu(ed[1].bitmap_start_clu);
```

Comment:
Shouldn't we check the type of the directory entry?

Reply:
As I said, I should show ondisk information even though it is corrupted.